src/seq/demultiplexer.jl

```diff
@@ -0,0 +1,336 @@
+        nodes[l] = 0
+    end
+    base = endof(nodes) - 5
+
+    # calculate ranges for the next level
+    starts = Vector{Int}(5)
+    i = first(range)
+    starts[1] = i
+    while i ≤ last(range) && length(barcodes[i]) < j
+        i += 1
+    end
+    for k in 1:4
+        starts[k+1] = i
+        while i ≤ last(range) && Int(barcodes[i][j]) + 1 == k
+            i += 1
+        end
+    end
+
+    # set nodes
+    nodes[root] = (base << 1) | 1
+    for k in 1:5
+        if k == 5
+            r = starts[k]:last(range)
+        else
+            r = starts[k]:starts[k+1]-1
+        end
+        build_trie!(nodes, base + k, barcodes, ids, r, j + 1)
+    end
+end
+
+# Find the longest match of `seq` prefixes in `trie`.
+function findbarcode(trie::BarcodeTrie, seq)
+    s = 1
+    for nt in seq
+        t = (trie.nodes[s] >> 1) + min(Int(nt), 3) + 2
+        if trie.nodes[t] & 1 == 0
+            # leaf
+            break
+        end
+        s = t
+    end
+    s = (trie.nodes[s] >> 1) + 1
+    return trie.nodes[s] >> 1
+end
+
+type Demultiplexer{S}
+    # barcode tries with ≥ 0 errors in ascending order
+    # (i.e. tries[1]: no errors, tries[2]: one error, ...)
+    tries::Vector{BarcodeTrie}
+
+    # original barcodes
+    barcodes::Vector{S}
+
+    # distance (:hamming or :levenshtein)
+    distance::Symbol
+end
+
+function Base.show(io::IO, demultiplexer::Demultiplexer)
+    println(io, summary(demultiplexer), ":")
+    println(io, "  distance: ", demultiplexer.distance)
+    println(io, "  number of barcodes:", length(demultiplexer.barcodes))
+      print(io, "  number of correctable errors: ", length(demultiplexer.tries) - 1)
+end
+
+"""
+    Demultiplexer(barcodes::Vector{DNASequence};
+                  n_max_errors::Integer=1,
+                  distance::Symbol=:hamming)
+
+Create a demultiplexer object from `barcodes`.
+
+# Arguments
+* `barcodes`: the sorted list of DNA barcodes; integer IDs are assigned in this order.
+* `n_max_errors=1`: the number of maximum correctable errors.
+* `distance=:hamming`: the distance metric (`:hamming` or `:levenshtein`).
+"""
+function Demultiplexer(barcodes::Vector{DNASequence};
+                       n_max_errors::Integer=1,
+                       distance::Symbol=:hamming)
+    if !issorted(barcodes)
+        error("barcodes must be sorted")
+    elseif n_max_errors < 0
+        error("n_max_errors must be non-negative")
+    elseif distance ∉ (:hamming, :levenshtein)
+        error("distance must be either :hamming or :levenshtein")
+    end
+
+    # check barcodes
+    for barcode in barcodes
+        for i in 1:endof(barcode)
+            if isambiguous(barcode[i])
+                error("barcode must be umambiguous")
+            end
+        end
+    end
+
+    # build trie trees for each error level (0-n_max_errors)
+    tries = BarcodeTrie[]
+    for m in 0:n_max_errors
+        # generate "erroneous" barcodes
+        barcodes′ = DNASequence[]
```

Comment:
personal pet peeve, but I'd like to `s/barcodes'/mutated_barcodes/` for clarity. Feel free to ignore this